p2p/protocol/identify/obsaddr.go

```diff
@@ -185,20 +230,92 @@ func (oas *ObservedAddrSet) gc() {
 	}
 }
 
-func (oas *ObservedAddrSet) doAdd(observed, local, observer ma.Multiaddr,
-	direction network.Direction) {
+func (oas *ObservedAddrManager) addConn(conn network.Conn, observed ma.Multiaddr) {
+	oas.activeConnsMu.Lock()
+	defer oas.activeConnsMu.Unlock()
+
+	// We need to make sure we haven't received a disconnect event for this
+	// connection yet. The only way to do that right now is to make sure the
+	// swarm still has the connection.
+	//
+	// Doing this under a lock that we _also_ take in a disconnect event
+	// handler ensures everything happens in the right order.
+	for _, c := range oas.host.Network().ConnsToPeer(conn.RemotePeer()) {
```

Comment:
warn / error if the conn isn't matched?

Reply:
That's an expected edge-case. It won't be matched if we disconnected while processing the address observation. This check ensures that we don't store the connection in that case, and leak memory.